text/3334-niche.md

```diff
@@ -0,0 +1,331 @@
+struct MeaninglessNumber(u64);
+
+assert_eq!(size_of::<MeaninglessNumber>(), 8);
+assert_eq!(size_of::<Option<MeaninglessNumber>>(), 8);
+
+#[niche(range = 2..)]
+struct Bit(u8);
+
+assert_eq!(size_of::<Bit>(), 1);
+assert_eq!(size_of::<Option<Option<Option<Bit>>>>(), 1);
+```
+
+Constructing a structure with a niche value, or writing to the non-ZST field of
+such a structure, or obtaining a mutable reference to such a field, requires
+`unsafe` code. Causing a type with a niche to contain an invalid value (whether
+by construction, writing, or transmuting) results in undefined behavior.
+
+If a type `T` contains only a single niche value, `Option<T>` (and other enums
+isomorphic to it, with one variant containing `T` and one nullary variant) will
+use that value to represent `None` (the nullary variant). If such a `T` is
+additionally `repr(transparent)` or `repr(C)` or otherwise permitted in FFI,
+`Option<T>` will likewise be permitted in FFI, with the niche value mapping
+bidirectionally to `None` across the FFI boundary.
+
+If a type contains multiple niche values, Rust does not guarantee any
+particular mapping at this time, but may in the future.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The attribute `#[niche]` may only appear on a struct declaration. The struct
+must contain exactly one field of a non-zero-sized type (non-ZST). The struct
+may contain zero or more ZST fields, such as `PhantomData`.
+
+Declaring a niche on any item other than a struct declaration results in an
+error.
+
+Declaring multiple `niche` attributes on a single item, or multiple key-value
+pairs within a single `niche` attribute, results in an error.
+
+Declaring a niche on a struct containing more or less than one non-zero-sized
+field results in an error.
+
+Declaring a niche on a struct that has any generic parameters affecting the
+non-zero-sized field results in an error.
+
+Declaring a range niche with an empty range (e.g. `0..0`) results in a
+warn-by-default lint. As with many lints, this lint should be automatically
+suppressed for code expanded from a macro.
+
+Declaring a range niche with an invalid range (e.g. `5..0`) results in an
+error.
+
+Declaring a niche using a negative value or a negative range endpoint results
+in an error. The representation of negative values depends on the size of the
+type, and the compiler may not have that information at the time it handles
+attributes such as `niche`. The text of the error should suggest the
+appropriate unsigned equivalent to use. The compiler may support this in the
+future.
+
+Declaring a range niche with an open start (`..3`) results in an error, for
+forwards-compatibility with support for negative values.
+
+Declaring a niche using a non-literal value (e.g. `usize::MAX`) results in an
+error. Constants can use compile-time evaluation, and compile-time evaluation
+does not occur early enough for attributes such as niche declarations.
```

Comment:
This is unfortunate, as I’d expect `#[niche(value = MAX)]` to be a common use case.

Perhaps this is an acceptable work-around:

```rust
const _: () = assert!(4294967295 == u32::MAX);
#[niche(value = 4294967295)]
struct Idx(u32);

Reply:
@SimonSapin I hope we can make it work in the future. cc @oli-obk

Reply:
Somehow I forgot about hex literals. This looks reasonable:

```rust
#[niche(value = 0xffff_ffff)]
struct Idx(u32);

Reply:
> Somehow I forgot about hex literals. This looks reasonable:
> 
> ```rust
> #[niche(value = 0xffff_ffff)]
> struct Idx(u32);
> ```

This pattern doesn't check for types (since literal types are automatically inferred). If someone changes `u32` to `u64` (especially through a type alias) you will easily forget to update this. Also `usize::MAX`?